### PR TITLE
Fix 4121and4125

### LINKED ISSUE #4121: Reset command in Interactive adds "reset \r\n" to history instead of "reset"
Repro:
1. Open C# interactive window and type #reset and hit enter
2. Now type ctrl + up arrow.

expected: We see only reset command in the history.
actual: we see "reset \r\n"

We need to trim the snapshot before adding it to history. This is causing it to add the same command twice because it thinks it is a different command than previous one because of \r\n in the end..

capture

### PR DESCRIPTION
**Pull request by ManishJayaswal**
The root issue here was the call to AddInput method ( which echoes the command on the language buffer) which was being made even when the reset command was typed at the prompt instead of just when reset button was clicked. The language buffer actually gets marked as readonly when input on the prompt gets submitted. We now check to see if the language buffer is readonly and bail out of AddInput if it is.  Previously AddInput was trying to echo the reset command which did nothing as the buffer was readonly but the subsequent processing in AddInput was causing all these issues. 

Fixes #4121, #4125 and #4277

### CODE REVIEW

**Comment by ManishJayaswal** [comment]
@tmat @amcasey @cston @KevinH-MS FYI.
This eliminates the need to add extra interfaces to distinguish between reset button click and reset command on the prompt. 
We would still need to add smart reset command for C# and VB interactive window that honors users reset arguments. I have files a separate issue ( #4503) to track that.

File: src/InteractiveWindow/Editor/InteractiveWindow_UIThread.cs [hunk 1]
```diff
@@ -334,6 +334,12 @@ private void PendSubmissions(IEnumerable<PendingSubmission> inputs)
 
             public void AddInput(string command)
             {
+                // If the language buffer is readonly then input can not be added. Return immediately.
+                if (_window._currentLanguageBuffer != null && _window._currentLanguageBuffer.IsReadOnly(0))
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow_UIThread.cs]: When does this occur?  Would it be appropriate to simply add a new, writeable language buffer, as when it's null?

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs [hunk 2]
```diff
@@ -629,12 +629,15 @@ public void CheckHistoryPrevious()
...
 
-        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/4121")]
-        public void CheckHistoryAfterResetButtonClick()
+        [Fact]
+        public void CheckHistoryPreviousAfterReset()
         {
-            Task.Run(() => Window.Operations.ResetAsync(initialize: true)).PumpingWait();
-            Task.Run(() => Window.Operations.HistoryPrevious()).PumpingWait();
-            Assert.Equal("#reset", GetTextFromCurrentLanguageBuffer);
+            const string V = "#reset";
```
Comment (amcasey) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: Name?

File: src/InteractiveWindow/Editor/InteractiveWindow_UIThread.cs (hunk 1 above)
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow_UIThread.cs]: Can you please add a comment about when this occurs?  Something like the description for this PR.

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs (hunk 2 above)
Comment (amcasey) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: Would it be worthwhile to check that there is only a single history entry for "#reset"?  Also, should there be a test of the history after using the reset button?

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs [hunk 3]
```diff
@@ -629,12 +629,15 @@ public void CheckHistoryPrevious()
...
+            Window.InsertCode(V);
+            Assert.Equal(V, GetTextFromCurrentLanguageBuffer);
+            Task.Run(() => Window.Operations.ExecuteInput()).PumpingWait();
+            Window.Operations.HistoryPrevious();
+            Assert.Equal(V, GetTextFromCurrentLanguageBuffer);
```
Comment (cston) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: Not part of this change, but 'GetTextFromCurrentLanguageBuffer' is an odd name for a property. Consider changing to a method.

**Comment by cston** [comment]
LGTM

File: src/InteractiveWindow/Editor/InteractiveWindow_UIThread.cs (hunk 1 above)
Comment (ManishJayaswal) [inline src/InteractiveWindow/Editor/InteractiveWindow_UIThread.cs]: Done

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs (hunk 3 above)
Comment (ManishJayaswal) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: changed it to a method

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs (hunk 2 above)
Comment (ManishJayaswal) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: Changed the name. Sorry I was just lazy and it was late :-)
Yes it would be worthwhile to test that history only has a single entry. I have a issue (4307) assigned to me to write more tests for these bugs. I will include this as well.
